python/pyspark/ml/tests/test_evaluation.py

```diff
@@ -14,18 +14,368 @@
+        )
+
+        dataset_with_weight = self.spark.createDataFrame(
+            featureAndPredictionsWithWeight, ["features", "prediction", "weight"]
+        )
+        evaluator.setWeightCol("weight")
+
+        # Evaluate the dataset with weights
+        score_with_weight = evaluator.evaluate(dataset_with_weight)
+        self.assertTrue(np.allclose(score_with_weight, 0.9079, atol=1e-4))
+
+        # read/write
+        with tempfile.TemporaryDirectory(prefix="save") as tmp_dir:
+            # Save the evaluator
+            ce_path = tmp_dir + "/ce"
+            evaluator.write().overwrite().save(ce_path)
+            # Load the saved evaluator
+            evaluator2 = ClusteringEvaluator.load(ce_path)
+            self.assertEqual(evaluator2.getPredictionCol(), "prediction")
+
+    def test_clustering_evaluator_with_cosine_distance(self):
+        featureAndPredictions = map(
+            lambda x: (Vectors.dense(x[0]), x[1]),
+            [
+                ([1.0, 1.0], 1.0),
+                ([10.0, 10.0], 1.0),
+                ([1.0, 0.5], 2.0),
+                ([10.0, 4.4], 2.0),
+                ([-1.0, 1.0], 3.0),
+                ([-100.0, 90.0], 3.0),
+            ],
+        )
+        dataset = self.spark.createDataFrame(featureAndPredictions, ["features", "prediction"])
+        evaluator = ClusteringEvaluator(predictionCol="prediction", distanceMeasure="cosine")
+        self.assertEqual(evaluator.getDistanceMeasure(), "cosine")
+        self.assertTrue(np.isclose(evaluator.evaluate(dataset), 0.992671213, atol=1e-5))
+
+    def test_regression_evaluator(self):
+        dataset = self.spark.createDataFrame(
+            [
+                (-28.98343821, -27.0),
+                (20.21491975, 21.5),
+                (-25.98418959, -22.0),
+                (30.69731842, 33.0),
+                (74.69283752, 71.0),
+            ],
+            ["raw", "label"],
+        )
+
+        evaluator = RegressionEvaluator()
+        evaluator.setPredictionCol("raw")
+
+        # Evaluate dataset with default metric (RMSE)
+        rmse = evaluator.evaluate(dataset)
+        self.assertTrue(np.allclose(rmse, 2.8424, atol=1e-4))
+        # Evaluate dataset with R2 metric
+        r2 = evaluator.evaluate(dataset, {evaluator.metricName: "r2"})
+        self.assertTrue(np.allclose(r2, 0.9939, atol=1e-4))
+        # Evaluate dataset with MAE metric
+        mae = evaluator.evaluate(dataset, {evaluator.metricName: "mae"})
+        self.assertTrue(np.allclose(mae, 2.6496, atol=1e-4))
+        # read/write
+        with tempfile.TemporaryDirectory(prefix="save") as tmp_dir:
+            # Save the evaluator
+            re_path = tmp_dir + "/re"
+            evaluator.write().overwrite().save(re_path)
+            # Load the saved evaluator
+            evaluator2 = RegressionEvaluator.load(re_path)
+            self.assertEqual(evaluator2.getPredictionCol(), "raw")
+
+        dataset_with_weights = self.spark.createDataFrame(
```

Comment:
Nit, I think for such cases, we can just build a dataset with weights, and then:
1, test without weightCol;
2, test with weightCol.

Just to save some time (our CI is becoming slower and slower)

Reply:
sure, all the test cases are from the example of the evaluators. I can refine it.

Reply:
Done.